Tests/Basic/SyncronizedQueueTests.swift

```diff
@@ -0,0 +1,84 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright 2015 - 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import XCTest
+import Utility
+import Basic
+
+class SyncronizedQueueTests: XCTestCase {
+
+    func testSingleProducerConsumer() {
+        let queue = SyncronizedQueue<Int>()
+        let queueElements = Set(0..<10)
+        var consumed = Set<Int>()
+
+        let producer = Thread {
+            for element in queueElements {
+                queue.enqueue(element)
+            }
+            queue.enqueue(nil)
+        }
+
+        let consumer = Thread {
+            while let element = queue.dequeue() {
+                consumed.insert(element)
+            }
+        }
+
+        consumer.start()
+        producer.start()
+        consumer.join()
+        producer.join()
+
+       XCTAssertEqual(consumed, queueElements)
+    }
+
+    func testMultipleProducerConsumer() {
+        let queue = SyncronizedQueue<Int>()
+
+        let queueElementsOne = Set(0..<10)
+        let queueElementsTwo = Set(11..<20)
+
+        var consumed = Set<Int>()
+
+        // Create two producers.
+        let producers = [queueElementsOne, queueElementsTwo].map { queueElements in
+            return Thread {
+                for element in queueElements {
+                    queue.enqueue(element)
+                }
+                queue.enqueue(nil)
+            }
+        }
+
+        // Create two consumers.
+        let consumers = [0, 1].map { _ in
+            return Thread {
+                while let element = queue.dequeue() {
+                    consumed.insert(element)
+                }
+            }
+        }
+
+        consumers.forEach { $0.start() }
+        producers.forEach { $0.start() }
+        // Block until all producers and consumers are done.
+        consumers.forEach { $0.join() }
+        producers.forEach { $0.join() }
+
+        // Make sure everything was consumed.
+        XCTAssertEqual(consumed, queueElementsOne.union(queueElementsTwo))
+    }
+
+    static var allTests = [
+        ("testSingleProducerConsumer", testSingleProducerConsumer),
+        ("testMultipleProducerConsumer", testMultipleProducerConsumer),
```

Comment:
It would be good to have at least one test that tried hard to exercise the signaling behavior of the queue. For example, making the producer add elements slowly (e.g. by locked access to the consumed set and waiting for it to an observe an insert before adding new items) so that the consumer is frequently going to be blocked.